File: 2021 포스코ICT/MSA 기초 교육/3 TypeScript/src/step4/da.map/MemberMapStore.ts
import CommunityMember from '../../step1/entity/club/CommunityMember';
import MemberStore from '../store/MemberStore';
import MemoryMap from './io/MemoryMap';

class MemberMapStore implements MemberStore {
    memberMap: Map<string, CommunityMember>;

    constructor() {
        this.memberMap = MemoryMap.getInstance().memberMap;
    }

    create(member: CommunityMember): string {
        const targetMember = this.memberMap.get(member.getId());

        if (targetMember) {
            throw new Error('Member already exists with email');
        }

        this.memberMap.set(member.getId(), member);

        return member.getId();
    }

    retrieve(email: string): CommunityMember | null {
        return this.memberMap.get(email) || null;
    }
    
    retrieveByName(name: string): CommunityMember[] {
        const members = Array.from(this.memberMap.values());
        return members.filter(member => member.name === name);
    }

    update(member: CommunityMember): void {
        this.memberMap.set(member.getId(), member);
    }

    delete(email: string): void {
        this.memberMap.delete(email);
    }

    exists(email: string): boolean {
        return this.memberMap.get(email) !== undefined;
    }
}
export default MemberMapStore;

File: 2021 포스코ICT/MSA 기초 교육/1 Java/src/step4/da/map/MemberMapStore.java
package step4.da.map;

import step1.entity.club.CommunityMember;
import step4.da.map.io.MemoryMap;
import step4.store.MemberStore;
import step4.util.MemberDuplicationException;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

public class MemberMapStore implements MemberStore {

    private Map<String, CommunityMember> memberMap;

    public MemberMapStore(){
        this.memberMap = MemoryMap.getInstance().getMemberMap();
    }


    @Override
    public String create(CommunityMember member) {
        Optional.ofNullable(memberMap.get(member.getId()))
                .ifPresent(entity -> {
                    throw new MemberDuplicationException("Member already exists with email: " + member.getId());
                });

        memberMap.put(member.getId(), member);
        return member.getId();
    }

    @Override
    public CommunityMember retrieve(String email) {
        return memberMap.get(email);
    }

    @Override
    public List<CommunityMember> retrieveByName(String name) {
        return memberMap.values().stream()
                .filter(member -> member.getName().equals(name))
                .collect(Collectors.toList());
    }

    @Override
    public void update(CommunityMember member) {
        memberMap.put(member.getId(), member);
    }

    @Override
    public void delete(String email) {
        memberMap.remove(email);
    }

    @Override
    public boolean exists(String email) {
        return Optional.ofNullable(memberMap.get(email)).isPresent();
    }
}
